### ISSUE TITLE
Unexpected TypeCoercion with int-literal param

### ISSUE THREAD

**Issue by SignpostMarv** [vimeo/psalm#1371]
# issue 
Passing known-matching literals from a loop whose bounds provably match the set of int-literals specified on a function parameter should not trigger `TypeCoercion` for `parent type int`

# example
* ~https://getpsalm.org/r/b74adb9f11~ https://getpsalm.org/r/3a77f3f0bf
* using 6b04523

> ERROR: TypeCoercion - 10:12 - Argument 1 of foo expects int(1)|int(2)|int(3)|int(4)|int(5)|int(6)|int(7)|int(8)|int(9)|int(10)|int(11)|int(12), parent type int provided
>
> ERROR: TypeCoercion - 14:12 - Argument 1 of foo expects int(1)|int(2)|int(3)|int(4)|int(5)|int(6)|int(7)|int(8)|int(9)|int(10)|int(11)|int(12), parent type int provided

```php
<?php
/**
* @psalm-param 1|2|3|4|5|6|7|8|9|10|11|12 $k
*/
function foo(int $k) : string {
  return (string) $k;
}

for ($i=1;$i<=12;++$i) {
  echo foo($i), "\n";
}

foreach (range(1, 12) as $i) {
  echo foo($i), "\n";
}
```
# expected
Under the following conditions, no error should be reported:
* if `$k` is not passed to any function or method that allows by-ref manipulation ***and***
* if `$i` originates from a `for(;;)` loop is not modified to exceed the limit *before* `foo()` is called ***and/or***
* a loop produces a range of values exactly matching the set of int-literals

## should fail
```php
function bar(int $limit) : void {
    for ($i=1;$i<=$limit;++$i) { // should fail because $limit is not "known"
        echo foo($i), "\n";
    }
}
```

## looks like it should fail, but can't
```php
foreach (range(1, 12, 7) as $i) { // range produces array{0:1, 1:8}, not array{0:1, 1:8, 2:15}
    echo foo($i), "\n";
}
```

# related
Ability to specify range literals may be handy- i.e. if it's valid to pass to [`range()`](http://php.net/manual/en/function.range.php), i.e. `@psalm-param 1...12 $k`, or for strings `@psalm-param 'a'...'i' $k`, `@psalm-param 'c'...'a' $k`

**Comment by muglug** [comment]
Not going to fix this, I don't think - this gets into a bit of a minefield of massive type definitions

**Comment by SignpostMarv** [comment]
nee probs, this is what the baseline file is for :)

**Comment by SignpostMarv** [comment]
> this gets into a bit of a minefield of massive type definitions

shall I file the range-literals request separately?

**Comment by muglug** [comment]
Better - PR it. You might want to cap it at something like 20 types? RangeReturnTypeProvider is your friend.
